src/scheduler/scheduler.ts

```diff
@@ -0,0 +1,183 @@
+            if (ro.schedulerConfig) {
+                logger.info(`    ${ro.name} @ ${ro.schedulerConfig.crontab ?? '* * * * *'}; ${ro.schedulerConfig.mode ?? 'Exactly Once Per Interval'}`);
+            }
+        });
+    }
+}
+
+class DetachableLoop {
+    private isRunning: boolean = false;
+    private resolveCompletion?: (value: void | PromiseLike<void>) => void;
+    private interruptResolve?: () => void;
+    private lastExec: Date;
+    private timeMatcher: TimeMatcher;
+
+    constructor(readonly dbosExec: DBOSExecutor, readonly crontab: string, readonly mtd: SchedulerRegistration<unknown, unknown[], unknown>) {
+        this.lastExec = new Date();
+        this.lastExec.setMilliseconds(0);
+        this.timeMatcher = new TimeMatcher(crontab);
+    }
+
+    async startLoop(): Promise<void> {
+        // See if the exec time is available in durable storage...
+        const lasttm = await this.dbosExec.systemDatabase.getLastScheduledTime(this.mtd.name);
+        if (lasttm) {
+            this.lastExec = new Date(lasttm);
+        }
+
+        this.isRunning = true;
+        while (this.isRunning) {
+            const nextExecTime = this.timeMatcher.nextWakeupTime(this.lastExec);
+            const sleepTime = nextExecTime.getTime() - new Date().getTime();
+            console.log(`Loop iteration with sleep time: ${sleepTime}ms`);
+
+            if (sleepTime > 0) {
+                // Wait for either the timeout or an interruption
+                await Promise.race([
+                    this.sleep(sleepTime),
+                    new Promise<void>((_, reject) => this.interruptResolve = reject)
+                ])
+                .catch(); // Interrupt sleep throws
+            }
+
+            if (!this.isRunning) {
+                break;
+            }
+
+            // Check crontab
+            if (!this.timeMatcher.match(nextExecTime)) {
+                this.lastExec = nextExecTime;
+                continue;
+            }
+
+            // Init workflow
+            const workflowUUID = `sched-${this.mtd.name}-${nextExecTime.toISOString()}`;
+            this.dbosExec.logger.debug(`Executing scheduled workflow ${workflowUUID}`);
+            const wfParams = { workflowUUID: workflowUUID };
+            // All operations annotated with Scheduled decorators must take in these four
+            const args: ScheduledArgs = [nextExecTime, new Date(), 0, 0]; // TODO calculate outstanding numbers
+
+            // We can only guarantee exactly-once-per-message execution of transactions and workflows.
+            if (this.mtd.workflowConfig) {
+                // Execute the transaction
+                await this.dbosExec.workflow(this.mtd.registeredFunction as Workflow<[Date, Date, number, number], unknown>, wfParams, ...args);
+            }
+            else {
+                this.dbosExec.logger.error(`Function ${this.mtd.name} is @scheduled but not a workflow`);
+            }
```

Comment:
Could we fail earlier, before we even start this loop?

Reply:
It should actually fail in the compiler, if you used it.  Is that early enough?

Reply:
Ha, ok. Well, why not remove the branch at all then? Or do it in the constructor?